src/htsjdk/java/htsjdk/samtools/BinningIndexBuilder2.java

```diff
@@ -0,0 +1,200 @@
+                lastChunk.setChunkEnd(chunkEnd);  // coalesced
+            } else {
+                oldChunks.add(newChunk);
+                bin.setLastChunk(newChunk);
+            }
+        }
+
+        // process linear index
+
+        // the smallest file offset that appears in the 16k window for this bin
+        final int featureEnd = feature.getEnd();
+        int startWindow = LinearIndex.convertToLinearIndexOffset(feature.getStart()); // the 16k window
+        final int endWindow;
+
+        if (featureEnd == GenomicIndexUtil.UNSET_GENOMIC_LOCATION) {   // assume feature uses one position
+            // Next line for C (samtools index) compatibility. Differs only when on a window boundary
+            startWindow = LinearIndex.convertToLinearIndexOffset(feature.getStart() - 1);
+            endWindow = startWindow;
+        } else {
+            endWindow = LinearIndex.convertToLinearIndexOffset(featureEnd);
+        }
+
+        if (endWindow > largestIndexSeen) {
+            largestIndexSeen = endWindow;
+        }
+
+        // Set the linear index at every 16K window that this feature overlaps, but only if this chunk
+        // start is strictly earlier than any previously seen chunk start for the window, or if the window
+        // is uninitialized (this is the first feature overlapping this window).
+        for (int win = startWindow; win <= endWindow; win++) {
+            // Initially each window is set to UNINITIALIZED_WINDOW (-1) so that we can distinguish here between
+            // windows that have no overlapping features, and those whose lowest feature offset is legitimately 0,
+            // which is a valid (virtual file) offset for feature formats that don't require a header.
+            if (index[win] == UNINITIALIZED_WINDOW || chunkStart < index[win]) {
+                index[win] = chunkStart;
+            }
+        }
+    }
+
+    /**
+     * Creates the BAMIndexContent for this reference.
+     * Requires all features of the reference have already been processed.
+     */
+    public BinningIndexContent generateIndexContent() {
+
+
+        // process bins
+        if (binsSeen == 0) return null;  // no bins for this reference
+
+        // process chunks
+        // nothing needed
+
+        // process linear index
+        // linear index will only be as long as the largest index seen
+        final long[] newIndex = new long[largestIndexSeen + 1]; // in java1.6 Arrays.copyOf(index, largestIndexSeen + 1);
+
+        // C (samtools index) also fills in intermediate 0's with values.  This seems unnecessary, but safe
+        long lastNonZeroOffset = 0;
+        for (int i = 0; i <= largestIndexSeen; i++) {
+            if (index[i] == UNINITIALIZED_WINDOW) {
+                // Following line removed so that we can distinguish alignment == 0 from no alignment.
+                // Note this is only for temporary intermediate BAI files.
```

Comment:
I'm not sure what impact it would have on other tools, so we shouldn't change it in htsjdk. We could make it an option that is exposed and only used by this codepath.